### PR TITLE
fix: `into_iterator` not completed on `Vec<{unknown}>`

### LINKED ISSUE #10297: `into_iterator` not resolved or completed on `Vec<{unknown}>`
It feels like we could offer `into_iter` here, since the impl is not constrained in any way:
```rust
fn main() {
    let mut x = Vec::new();
    x.into_iter();
}

// impl<T, A: Allocator> IntoIterator for Vec<T, A>
```

### PR DESCRIPTION
**Pull request by iDawer**
Fixes  #10297

### CODE REVIEW

**Comment by lnicola** [comment]
Ah, I didn't manage to write a test here.

Should we do the same thing in `autoderef`?

**Comment by iDawer** [comment]
Actually there are more: `impls_future`, `impls_fnonce`, `impls_trait`. Do we need tests for them? Let me see if there are trivial tests

**Comment by iDawer** [comment]
I couldn't reproduce this in tests but RA definetely fails on it (`hir::Type::autoderef` fixed):

```rust
struct Foo<T>;
struct Bar;
trait Trait {
    fn quux(&self) {}
}
impl<T> core::ops::Deref for Foo<T> {
    type Target = Bar;
    fn deref(&self) -> &Self::Target;
}
impl Trait for Bar {}
fn main() {
    let x: Foo<_>;
    x.$0
    // missing `quux` method in completions
}
```

---
I had tried to fix `impls_trait` but Chalk started panicking on the next example. This needs some elaboration.

```rust
struct Vec<T>;
impl<T> IntoIterator for Vec<T> {
    type Item = ();
    type IntoIter = ();
    fn into_iter(self);
}
fn main() {
    let x: Vec<_>;
    x.$0;
    // missing postfix completion "for"
}
```
I'll leave this and others for next time

File: crates/hir/src/lib.rs [hunk 1]
```diff
@@ -2505,8 +2505,8 @@ impl Type {
     pub fn autoderef<'a>(&'a self, db: &'a dyn HirDatabase) -> impl Iterator<Item = Type> + 'a {
         // There should be no inference vars in types passed here
         // FIXME check that?
```
Comment (flodiebold) [inline crates/hir/src/lib.rs:2507]: this FIXME is also fixed by this

File: crates/hir/src/lib.rs [hunk 2]
```diff
@@ -2601,9 +2601,7 @@ impl Type {
     ) {
         // There should be no inference vars in types passed here
         // FIXME check that?
```
Comment (flodiebold) [inline crates/hir/src/lib.rs:2603]: as above

**Review by flodiebold — APPROVED** [review]
bors d+

**Comment by iDawer** [comment]
This PR caused regression in nightly  #10396

**Pull request merged**